File: firmware/plane/src/effects.c
#include "effects.h"

#include <math.h>
#include <stdbool.h>
#include <string.h>

#include "config.h"

typedef struct {
    uint16_t index;
    uint8_t r, g, b;
} nav_light_t;

#if NAV_COUNT > 0
static const nav_light_t s_nav[] = NAV_LIGHTS;
#endif
static uint8_t s_gamma[256];
static uint32_t s_rand = 0x12345678;

// ---------------------------------------------------------------- helpers ---

static uint8_t xorshift(void) {
    s_rand ^= s_rand << 13;
    s_rand ^= s_rand >> 17;
    s_rand ^= s_rand << 5;
    return (uint8_t)(s_rand >> 16);
}

// Integer HSV with full saturation; hue 0..255 walks the colour wheel.
static rgb_t hsv(uint8_t hue, uint8_t value) {
    uint8_t sector = hue / 43;
    uint8_t offset = (uint8_t)((hue - sector * 43) * 6);
    uint8_t p = 0;
    uint8_t q = (uint8_t)(((uint16_t)value * (255 - offset)) >> 8);
    uint8_t t = (uint8_t)(((uint16_t)value * offset) >> 8);

    switch (sector) {
    case 0:  return (rgb_t){value, t, p};
    case 1:  return (rgb_t){q, value, p};
    case 2:  return (rgb_t){p, value, t};
    case 3:  return (rgb_t){p, q, value};
    case 4:  return (rgb_t){t, p, value};
    default: return (rgb_t){value, p, q};
    }
}

static inline uint8_t scale8(uint8_t value, uint8_t scale) {
    return (uint8_t)(((uint16_t)value * (uint16_t)scale) >> 8);
}

static void fill(rgb_t *pixels, uint16_t count, rgb_t colour) {
    for (uint16_t i = 0; i < count; i++) pixels[i] = colour;
}

// The show dimmer, the safety ceiling and gamma correction, applied once at the
// end so every effect gets the same treatment.
static void finish(rgb_t *pixels, uint16_t count, uint8_t brightness) {
    uint8_t level = scale8(brightness, MAX_BRIGHTNESS);
    for (uint16_t i = 0; i < count; i++) {
        pixels[i].r = s_gamma[scale8(pixels[i].r, level)];
        pixels[i].g = s_gamma[scale8(pixels[i].g, level)];
        pixels[i].b = s_gamma[scale8(pixels[i].b, level)];
    }
}

// param 0..255 maps to a 2000 ms .. 100 ms cycle.
static uint32_t period_ms(uint8_t param) {
    return 2000u - ((uint32_t)param * 1900u) / 255u;
}

// Position inside the current cycle, 0..255.
static uint8_t phase(uint32_t now_ms, uint32_t period) {
    return (uint8_t)(((now_ms % period) * 256u) / period);
}

// Triangle wave, so fades go up and back down smoothly.
static uint8_t triangle(uint8_t x) {
    return x < 128 ? (uint8_t)(x * 2) : (uint8_t)((255 - x) * 2);
}

// ---------------------------------------------------------------- effects ---

#define EFFECT_COUNT 11

void effects_render(const show_state_t *state, uint32_t now_ms,
                    rgb_t *pixels, uint16_t count) {
    if (state->cue == 0 || state->brightness == 0) {
        memset(pixels, 0, count * sizeof(rgb_t));
        return;
    }

    const uint32_t period = period_ms(state->param);
    const uint8_t ph = phase(now_ms, period);
    const rgb_t base = hsv(state->hue, 255);
    uint8_t effect = state->cue;
    if (effect >= EFFECT_COUNT) effect = 1;  // unknown cue: steady colour

    switch (effect) {
    case 1:  // solid
        fill(pixels, count, base);
        break;

    case 2: {  // breathe
        uint8_t level = triangle(ph);
        fill(pixels, count, (rgb_t){scale8(base.r, level), scale8(base.g, level),
                                    scale8(base.b, level)});
        break;
    }

    case 3:  // strobe: short flash at the start of each cycle
        fill(pixels, count, ph < 24 ? base : (rgb_t){0, 0, 0});
        break;

    case 4:  // double strobe
        fill(pixels, count,
             (ph < 16 || (ph >= 40 && ph < 56)) ? base : (rgb_t){0, 0, 0});
        break;

    case 5: {  // chase: a lit block running along the strip
        uint16_t head = (uint16_t)(((uint32_t)ph * count) >> 8);
        uint16_t width = count / 6 + 1;
        for (uint16_t i = 0; i < count; i++) {
            uint16_t distance = (uint16_t)((i + count - head) % count);
            pixels[i] = distance < width ? base : (rgb_t){0, 0, 0};
        }
        break;
    }

    case 6: {  // comet with a fading tail
        uint16_t head = (uint16_t)(((uint32_t)ph * count) >> 8);
        for (uint16_t i = 0; i < count; i++) {
            uint16_t distance = (uint16_t)((i + count - head) % count);
            uint8_t level = distance > 255 ? 0 : (uint8_t)(255 - distance * 255 / count);
            level = scale8(level, level);  // squared falloff, reads as a tail
            pixels[i] = (rgb_t){scale8(base.r, level), scale8(base.g, level),
                                scale8(base.b, level)};
        }
        break;
    }

    case 7:  // sparkle
        for (uint16_t i = 0; i < count; i++) {
            pixels[i] = (xorshift() < 24) ? base : (rgb_t){0, 0, 0};
        }
        break;

    case 8:  // rainbow rotating around the airframe
        for (uint16_t i = 0; i < count; i++) {
            pixels[i] = hsv((uint8_t)(state->hue + ph + i * 256 / count), 255);
        }
        break;

    case 9:  // police: alternating halves, hue ignored
        for (uint16_t i = 0; i < count; i++) {
            bool front = i < count / 2;
            bool first = ph < 128;
            pixels[i] = (front == first) ? (rgb_t){255, 0, 0} : (rgb_t){0, 0, 255};
        }
        break;

    case 10: {  // theater chase, every third pixel
        uint8_t step = (uint8_t)(((uint32_t)ph * 3) >> 8);
        for (uint16_t i = 0; i < count; i++) {
            pixels[i] = (i % 3 == step) ? base : (rgb_t){0, 0, 0};
        }
        break;
    }

    default:
        fill(pixels, count, base);
        break;
    }

    finish(pixels, count, state->brightness);
}

void effects_render_failsafe(uint32_t now_ms, rgb_t *pixels, uint16_t count) {
    // Slow amber pulse at a fixed, low level: unmistakably not part of the show
    // and still bright enough to find the model.
    uint8_t level = triangle(phase(now_ms, 2000));
    rgb_t colour = {scale8(255, level), scale8(120, level), 0};
    fill(pixels, count, colour);
    finish(pixels, count, 96);
}

void effects_apply_nav(rgb_t *pixels, uint16_t count) {
#if NAV_COUNT > 0
    for (uint16_t i = 0; i < sizeof(s_nav) / sizeof(s_nav[0]); i++) {
        if (s_nav[i].index >= count) continue;
        pixels[s_nav[i].index] = (rgb_t){
            s_gamma[scale8(s_nav[i].r, MAX_BRIGHTNESS)],
            s_gamma[scale8(s_nav[i].g, MAX_BRIGHTNESS)],
            s_gamma[scale8(s_nav[i].b, MAX_BRIGHTNESS)],
        };
    }
#else
    (void)pixels;
    (void)count;
#endif
}

uint8_t effects_count(void) { return EFFECT_COUNT; }

void effects_init(void) {
    // Gamma 2.2 so a linear dimmer channel looks linear to the eye. Built once
    // at boot; everything afterwards is a table lookup.
    for (int i = 0; i < 256; i++) {
        float normalised = powf((float)i / 255.0f, 2.2f);
        s_gamma[i] = (uint8_t)(normalised * 255.0f + 0.5f);
    }
}

File: host/lightshow/web/effects.js
/* The airborne effect engine, ported to the browser.
 *
 * This is a line by line port of firmware/plane/src/effects.c -- same integer
 * arithmetic, same cue numbers, same phase maths. It exists so the interface
 * can show the pattern an aircraft will actually fly instead of asking anyone
 * to imagine "cue 6 at hue 200".
 *
 * Because it is a copy, it can drift. tools/ctest/effect_dump.c compiles the
 * real C and host/tests/test_effects_preview.py compares the two pixel for
 * pixel, so drift fails the test suite rather than the show.
 *
 * No DOM, no imports: a plain script that defines globals, loadable in node
 * for exactly that test.
 */

'use strict';

const EFFECT_COUNT = 11;

/* Gamma 2.2, the same table effects_init() builds.
 *
 * The preview does NOT use this by default, and that is deliberate. The table
 * exists because LEDs are linear and eyes are not, so a gamma corrected LED
 * *looks* like the linear value. A screen already applies its own correction,
 * so running the values through it a second time would only make the preview
 * darker than the real aircraft. The cross-check switches it on to compare
 * against the firmware byte for byte. */
const GAMMA = new Uint8Array(256);
for (let i = 0; i < 256; i++) GAMMA[i] = Math.floor((i / 255) ** 2.2 * 255 + 0.5);

/** Integer HSV with full saturation; hue walks the colour wheel in 0..255. */
function hsv(hue, value) {
  const sector = Math.floor(hue / 43);
  const offset = (hue - sector * 43) * 6;
  const q = (value * (255 - offset)) >> 8;
  const t = (value * offset) >> 8;
  switch (sector) {
    case 0:  return [value, t, 0];
    case 1:  return [q, value, 0];
    case 2:  return [0, value, t];
    case 3:  return [0, q, value];
    case 4:  return [t, 0, value];
    default: return [value, 0, q];
  }
}

const scale8 = (value, factor) => (value * factor) >> 8;

/** param 0..255 maps to a 2000 ms .. 100 ms cycle. */
const periodMs = (param) => 2000 - Math.floor(param * 1900 / 255);

/** Position inside the current cycle, 0..255. */
const phaseOf = (nowMs, period) => Math.floor((nowMs % period) * 256 / period);

/** Triangle wave, so fades go up and back down smoothly. */
const triangle = (x) => x < 128 ? x * 2 : (255 - x) * 2;

let randomState = 0x12345678;

/** The firmware's xorshift, bit for bit. */
function xorshift() {
  randomState ^= randomState << 13;
  randomState ^= randomState >>> 17;
  randomState ^= randomState << 5;
  randomState >>>= 0;
  return (randomState >>> 16) & 0xff;
}

/** Restarts the sparkle sequence. Only the cross-check needs this. */
function resetRandom(seed = 0x12345678) { randomState = seed >>> 0; }

/**
 * Renders one zone into a flat RGB array.
 *
 * `show` is {cue, hue, brightness, param}, 0..255 each, exactly as the four RC
 * channels carry them. Options: `max` is MAX_BRIGHTNESS, `out` reuses a buffer,
 * `gamma` applies the firmware's correction (see the note on GAMMA above).
 */
function renderEffect(show, nowMs, count, options = {}) {
  const max = options.max ?? 255;
  const pixels = options.out && options.out.length === count * 3
    ? options.out : new Uint8ClampedArray(count * 3);

  if (!show.cue || !show.brightness) {
    pixels.fill(0);
    return pixels;
  }

  const period = periodMs(show.param);
  const ph = phaseOf(nowMs, period);
  const base = hsv(show.hue, 255);
  const effect = show.cue >= EFFECT_COUNT ? 1 : show.cue;

  const put = (index, rgb) => {
    pixels[index * 3] = rgb[0];
    pixels[index * 3 + 1] = rgb[1];
    pixels[index * 3 + 2] = rgb[2];
  };
  const scaled = (rgb, level) =>
    [scale8(rgb[0], level), scale8(rgb[1], level), scale8(rgb[2], level)];
  const BLACK = [0, 0, 0];

  switch (effect) {
    case 1:                                            // solid
      for (let i = 0; i < count; i++) put(i, base);
      break;

    case 2: {                                          // breathe
      const colour = scaled(base, triangle(ph));
      for (let i = 0; i < count; i++) put(i, colour);
      break;
    }

    case 3: {                                          // strobe
      const on = ph < 24 ? base : BLACK;
      for (let i = 0; i < count; i++) put(i, on);
      break;
    }

    case 4: {                                          // double strobe
      const on = (ph < 16 || (ph >= 40 && ph < 56)) ? base : BLACK;
      for (let i = 0; i < count; i++) put(i, on);
      break;
    }

    case 5: {                                          // chase
      const head = (ph * count) >> 8;
      const width = Math.floor(count / 6) + 1;
      for (let i = 0; i < count; i++)
        put(i, ((i + count - head) % count) < width ? base : BLACK);
      break;
    }

    case 6: {                                          // comet with a tail
      const head = (ph * count) >> 8;
      for (let i = 0; i < count; i++) {
        const distance = (i + count - head) % count;
        let level = distance > 255 ? 0 : 255 - Math.floor(distance * 255 / count);
        level = scale8(level, level);
        put(i, scaled(base, level));
      }
      break;
    }

    case 7:                                            // sparkle
      for (let i = 0; i < count; i++) put(i, xorshift() < 24 ? base : BLACK);
      break;

    case 8:                                            // rainbow
      for (let i = 0; i < count; i++)
        put(i, hsv((show.hue + ph + Math.floor(i * 256 / count)) & 0xff, 255));
      break;

    case 9:                                            // police, hue ignored
      for (let i = 0; i < count; i++) {
        const front = i < Math.floor(count / 2);
        put(i, (front === (ph < 128)) ? [255, 0, 0] : [0, 0, 255]);
      }
      break;

    case 10: {                                         // theater chase
      const step = (ph * 3) >> 8;
      for (let i = 0; i < count; i++) put(i, i % 3 === step ? base : BLACK);
      break;
    }

    default:
      for (let i = 0; i < count; i++) put(i, base);
      break;
  }

  // finish(): the show dimmer, the safety ceiling, and gamma last.
  const level = scale8(show.brightness, max);
  for (let i = 0; i < pixels.length; i++) {
    const value = scale8(pixels[i], level);
    pixels[i] = options.gamma ? GAMMA[value] : value;
  }
  return pixels;
}

/** The slow amber pulse that means "no RC link". No cue produces this colour. */
function renderFailsafe(nowMs, count, options = {}) {
  const max = options.max ?? 255;
  const pixels = options.out && options.out.length === count * 3
    ? options.out : new Uint8ClampedArray(count * 3);

  const pulse = triangle(phaseOf(nowMs, 2000));
  const colour = [scale8(255, pulse), scale8(120, pulse), 0];
  // finish() runs on the failsafe pattern too, so the fixed level of 96 is
  // still scaled by the model's ceiling before it reaches the LEDs.
  const level = scale8(96, max);

  for (let i = 0; i < count; i++) {
    for (let c = 0; c < 3; c++) {
      const value = scale8(colour[c], level);
      pixels[i * 3 + c] = options.gamma ? GAMMA[value] : value;
    }
  }
  return pixels;
}

